app/code/Magento/Catalog/etc/db_schema.xml

```diff
@@ -25,12 +25,12 @@
         <constraint xsi:type="primary" referenceId="PRIMARY">
             <column name="entity_id"/>
         </constraint>
+        <constraint xsi:type="unique" referenceId="CATALOG_PRODUCT_ENTITY_SKU">
+            <column name="sku"/>
```

Comment:
I don't think it's enough to change just schema. In case there were duplicate SKUs in the database what error will be shown?

Let's check if the issue is valid first.

Reply:
@orlangur In case there were duplicates the database would already be corrupt.
It should not exist and on instances where this has happend the corruption has yet to be know to the application.
I think that only changing the database schema is a bit rough but never implementing a constraint in mysql but only in php will create buggy instances.
All the PHP checks can be removed if the database has been correctly designed.
PHP can validate the data but should not enforce constraints on the data, this should be a task for the data storage service.